### PR TITLE
Add session option to Strapi query for Mongo Multi-Document transaction

### PR DESCRIPTION
**Pull request by iicdii**
This is a mirror of #5711 which did not pass the test.

#### Description of what you did:
1) I added optional object as the last parameter of the Strapi Query, so that you can pass extra values inside Strapi query.
```js
// As-is
strapi.query('post').create({ title: 'test' });
strapi.query('post').find({ title: 'test' }, []);

// To-be
strapi.query('post').create({ title: 'test' }, {});
strapi.query('post').find({ title: 'test' }, [], {});
```
2) I made sure to use the **session** value for all mongoose operations, which is necessary for Mongo Multi-Document transaction.
It does not affect to previous operations though, because all mongoose operations has "session" option and calling session(null) removes the session from the query. I set the default value for session to null.

#### Example of usage:
```js
const input = [{ title: 'test1' }, { title: 'test2' }, { title: 'test3' }];

const session = await strapi.query('post').model.startSession();
session.startTransaction();
try {
  for (let data of input) {
    await strapi.query('post').create(data, { session });
  }

  await session.commitTransaction();
  session.endSession();
} catch (e) {
  await session.abortTransaction();
  session.endSession();
  throw e;
}
```

### CODE REVIEW

**Comment by iicdii** [comment]
@alexandrebodin I made new fresh PR but it keeps failing on Mysql testing which is not relevant with this PR. It seems like happening to other PRs too.. 😕

**Comment by derrickmehaffy** [comment]
> @alexandrebodin I made new fresh PR but it keeps failing on Mysql testing which is not relevant with this PR. It seems like happening to other PRs too.. 😕

You can safely ignore this for the moment.

File: packages/strapi-connector-mongoose/lib/buildQuery.js [hunk 1]
```diff
@@ -79,14 +79,15 @@ const buildQuery = ({
   searchParam,
   populate = [],
   aggregate = false,
+  session = null,
 } = {}) => {
   const search = buildSearchOr(model, searchParam);
 
   if (!hasDeepFilters(filters.where) && aggregate === false) {
-    return buildSimpleQuery({ model, filters, search, populate });
+    return buildSimpleQuery({ model, filters, search, populate, session });
```
Comment (alexandrebodin) [inline packages/strapi-connector-mongoose/lib/buildQuery.js:87]: Can you pass with session in a second argument buildSImpleQuery({}, { session }) to differentiate parameter and query context. 

You can do the same for the other functions :)

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
Nice addition !

File: packages/strapi-connector-mongoose/lib/buildQuery.js (hunk 1 above)
Comment (iicdii) [reply packages/strapi-connector-mongoose/lib/buildQuery.js]: Sure, Done!

**Review by Convly — COMMENTED** [review]
Hi, I tried to test your work and didn't succeed to execute your code example.

I got the following error while testing with the default strapi mongo database (the one from the [docker-compose.yml](https://github.com/strapi/strapi/blob/master/docker-compose.dev.yml)).

`MongoError: This MongoDB deployment does not support retryable writes. Please add retryWrites=false to your connection string. `

Did you configure anything particular in your `database.js` file or on your database setup and forgot to mention?

**Comment by iicdii** [comment]
@Convly 
You need to config replica set to enable transaction in Mongo. Based on [docker-compose.yml](https://github.com/strapi/strapi/blob/master/docker-compose.dev.yml), run command below.
```sh
$> docker-compose up -d mongo
$> docker-compose exec mongo mongo mongodb://localhost:27017 -u root -p strapi --eval "rs.initiate()"
```

You can check the requirements for MongoDB transactions on the official website. [Link](https://docs.mongodb.com/manual/core/transactions/)
> For transactions on MongoDB 4.2 deployments (replica sets and sharded clusters), clients must use MongoDB drivers updated for MongoDB 4.2.

> In version 4.0, MongoDB supports multi-document transactions on replica sets.
In version 4.2, MongoDB introduces distributed transactions, which adds support for multi-document transactions on sharded clusters and incorporates the existing support for multi-document transactions on replica sets.

**Review by Convly — COMMENTED** [review]
Works well on my side, nice addition!
I'm pinging @alexandrebodin to see if he has anything to say about the implementation.
Also, could you update your branch with master, please?

File: packages/strapi-connector-mongoose/lib/queries.js [hunk 2]
```diff
@@ -435,20 +448,23 @@ module.exports = ({ model, strapi }) => {
     }
 
     // Create entry with no-relational data.
-    const entry = await model.create(data);
-
+    // When specify options, data must be an array.
+    const [entry] = await model.create([data], { session });
```
Comment (alexandrebodin) [inline packages/strapi-connector-mongoose/lib/queries.js:452]: Can you clarify the error message to mention it is a mongoose requirement to call the method with an array if we pass a second argument to create so we get more context on this strange behavior :)

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/strapi-connector-mongoose/lib/queries.js (hunk 2 above)
Comment (iicdii) [reply packages/strapi-connector-mongoose/lib/queries.js:456]: Make sense. I added extra explanation.

**Review by alexandrebodin — APPROVED** [review]
LGTM

**Comment by alexandrebodin** [comment]
Thank you @iicdii 💯

**Pull request merged**